### ISSUE TITLE
want log of actions taken through the API

### ISSUE THREAD

**Issue by davepacheco** [oxidecomputer/omicron#2236]
I think for MVP we're going to want a log of all actions taken via the API.  This will be invaluable for fielding support requests ranging from the most basic (user confusion) to more complicated (that are likely to reflect bugs in our system).

This is sometimes called an audit log.  I'm not sure if that implies other properties (like authentication or some other kind of verifiability) that aren't strictly needed for MVP.

We may be able to use the existing Nexus log for this, but I think we'll need more details than what's there now.  I think we'll basically want request bodies and response bodies, excluding particularly sensitive data where necessary.  This will probably have security/privacy considerations as well.

Another angle would be to log database queries for all write operations.  We may want both.

**Comment by david-crespo** [comment]
We will likely want to expose some kind of view on this log for operators. That might not be an MVP feature, but it will be smart to build it with that in mind. Also possibly filtering by silo, project, etc.

**Comment by plotnick** [comment]
I implemented something like this for Permission Slip (aka "the signing service"): [log.rs](https://github.com/oxidecomputer/permission-slip/blob/main/server/src/log.rs). It's pretty basic right now and probably full of bugs, but if it'd be useful, I could probably factor it out into a standalone crate pretty easily.

**Comment by david-crespo** [comment]
That sounds suspiciously like a blockchain. 🧐

**Comment by plotnick** [comment]
Indeed, or Git. Merkle trees FTW!

**Comment by askfongjojo** [comment]
For MVP timeframe, perhaps we can at least capture the login user who created the resource (e.g., instance, image) so that operators or project members can go after the right people when they need certain resources cleaned up. (I was going to proposed capturing "updated_by" and "run_state_updated_by" but those are marginally useful compared to a complete audit log.) If this makes sense, I can raise a separate issue to capture created_by everywhere.

**Comment by zephraph** [comment]
@askfongjojo I think having `created_by` as a standalone issue makes sense. @plotnick how complete is `log.rs`? Are there some examples of how it integrates so I can get a better sense of its usage?

**Comment by plotnick** [comment]
> @plotnick how complete is `log.rs`?

Right now it's _very_ simple, and includes only the most basic functionality: append an entry to the log and update a head pointer, fetch a list of entries from a given head back to a root, and verify such a list of entries. It currently has no facilities for querying/searching other than "fetch the entry that has this id and its history chain". It also could very much use a second (or more) pair of :eyes:, as I think I'm the only one that's really looked at it or used it.

>  Are there some examples of how it integrates so I can get a better sense of its usage?

So, the tests in `log.rs` exercise its basic capabilities (enumerated above). It's used for real in `server.rs` to log the internal representations of every request and reply: see [Server::request](https://github.com/oxidecomputer/permission-slip/blob/6835e561dc4e332c124ce9b8f1d512367320eb71/server/src/server.rs#L431) and [Server::log](https://github.com/oxidecomputer/permission-slip/blob/6835e561dc4e332c124ce9b8f1d512367320eb71/server/src/server.rs#L393).

**Comment by zephraph** [comment]
I chatted a bit to @plotnick about this offline. We need to do a small RFD about this but I think perhaps something simpler than log.rs would be suitable for MVP.

**Comment by davepacheco** [comment]
Maybe it's a useful stopgap, but `created_by` doesn't seem like the right thing to me.  It sounds a lot like "owner", which we've generally avoided along the idea that most organizations want shared ownership of resources to avoid a small lottery factor ("bus factor").  "created_by" is worse than "owner" because it can't be changed, although maybe it's better if it's presented as a simple fact ("this was created by so-and-so") and not an implication that that person is still in any way responsible for it.

If we want to communicate who to chase down about a resource, might a more useful heuristic be "who has privileges on this resource"?  When we have metering, "who pays for this resource" might be good too, but I assume that's post-MVP too.

I'm not sure it's out of the question to do something a little richer?  I'm imagining something like a `resource_audit` table storing a timestamp, actor, resource type and id, action, and maybe comment; and we insert into this table whenever we do anything.  There's a lot to be figured out there, but I wonder if something like that might be plausible for MVP.

**Comment by zephraph** [comment]
Yeah @davepacheco, the `resource_audit` table is something along the lines of what I was thinking. I'm creating a small RFD on it.

**Comment by davepacheco** [comment]
Cool.  I think it would be particularly useful to write down what we want from this feature.  Personally I'm interested in two things:

1. a way for a customer to debug or troubleshoot why something looks the way it does ("I thought I created this instance but I can't find it.  what gives?  oh, I can see in the project log that I deleted it yesterday.")
2. a way for Oxide support to reliably understand the sequence of user actions taken (which is potentially different from what the system actually did)

The thing I suggested is useful for some kinds of problems, but doesn't help with a lot of kinds of request failures because there won't be an audit log entry if no action was taken.  In that case, it would be more useful to have request/response bodies (excluding sensitive information) for all requests.  And if we had that, we might not need this other audit log.

There's also a separate use case around security auditing that I don't know much about.  I imagine it involves assessing who has accessed what resources (including maybe read actions? which I was not assuming would be in this log).  This has pretty different requirements if, say, we want to fail an operation if we cannot write the audit log entry.

I don't want to make a mess of use cases here and I don't have a concrete suggestion except to write down what problems we're looking to solve (and maybe which ones we're not).

**Comment by david-crespo** [comment]
Closing in favor of #8810 now that we have an MVP merged and are iterating.
